Erlay: bandwidth-efficient transaction relay protocol

naumenkogs opened the pull request:
This is an implementation of [Erlay](https://arxiv.org/abs/1905.10518), using primitives in the BIP-330 (see the updated spec [here](https://github.com/naumenkogs/bips/blob/bip_0330_updates/bip-0330.mediawiki)). Please refer to these two to understand the design. My talk on the topic is [here](https://youtu.be/YxsjdIl0034?t=670).

Erlay uses both flooding (announcing using INV messages to all peers) and reconciliation to announce transactions. Flooding is expensive, so Erlay seeks to use it sparingly and in strategic locations - only well-connected publicly reachable nodes flood transactions to other publicly reachable nodes via outbound connections. Since every unreachable node is directly connected to several reachable nodes, this policy ensures that a transaction is quickly propagated to be within one hop from most of the nodes in the network.

All transactions not propagated through flooding are propagated through efficient set reconciliation. To do this, every node keeps a reconciliation set for each peer, in which transactions are placed which would have been announced using INV messages absent this protocol. Every 2 seconds every node chooses a peer from its outbound connections in a predetermined order to reconcile with, resulting in both sides learning the transactions known to the other side. After every reconciliation round, the corresponding reconciliation set is cleared.

I think both paper and the BIP motives the changes, but I'll mention them briefly once again here:
- save 40% of the bandwidth consumed by a node
- increase network connectivity for almost no bandwidth or latency cost
- improves privacy as a side-effect

Obviously looking for review, let's try to start with a high-level concerns, and keep nits for later.

P.S.
Please don't be scared of 8,000 LOC added. 7,000 of them is minisketch added as a subtree.

P.P.S.
My experiments of running this code live (slightly outdated) with a script to replicate the experiment: [here1](https://github.com/bitcoin/bitcoin/pull/18261#issuecomment-599241490) and [here2](https://github.com/bitcoin/bitcoin/pull/18261#issuecomment-600388543).

naumenkogs committed:
Add minisketch dependency

luke-jr commented:
How does it react to diverse network policies?

Empact commented:
Concept ACK

practicalswift commented on src/net_processing.cpp:
```diff
@@ -3258,6 +3403,331 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
+
+        std::string salt1, salt2;
+        if (remote_salt < local_salt) {
+            salt1 = std::to_string(remote_salt);
```
I'm afraid `std::to_string` is locale dependent.

---

For those interested in permanently killing the locale dependency bug class, consider reviewing:
* #18124 – init: Clarify C and C++ locale assumptions. Add locale sanity checks to verify assumptions at run-time
* #18126 – tests: Add fuzzing harness testing the locale independence of the `strencodings.h` functions
* #18147 – qt: Kill the locale dependency bug class by not allowing Qt to mess with `LC_NUMERIC`

practicalswift reviewed (commented):
Concept ACK

Thanks for doing this! Excellent work!

naumenkogs committed:
Add tests for set reconciliation

naumenkogs commented:
@luke-jr 

When it comes to network policies, I'm using the same code originally used by regular gossip ("Determine transactions to relay" in net_processing.cpp). So nothing should be lost or sent wastefully as a result of policy discrepancies.

gmaxwell commented:
@luke-jr  I'm understanding your question as being inspired by earlier 'mempool sync' ideas that would bleed bandwidth if there was a long lasting mempool discrepancy.

Erlay isn't a mempool sync. It's uses a way of communicating lists of things you want to relay which only takes bandwidth proportional to the difference rather than the total size.  So there is no on-going bandwidth usage due to differences in *mempool* content. Bandwidth is used is roughly  A*ntx_relayed + B*peers*(difference_in_tx_relayed) + C*peers.  for some constants A,B,C. 

If a peer has a radically different relay policy than you, it works fine and continues to save bandwidth below what usage would be without erlay even though the erlay savings itself comes largely from eliminating data that both sides would send.

Does that answer your question?

practicalswift commented:
@naumenkogs When trying out this PR I ran in to two small testing issues:

* The suffix of the functional test `p2p_erlay` is `.p2p` (`p2p_erlay.p2p`) instead of the expected `.py` (`p2p_erlay.p2p`) :)
* It seems like `make check` runs the `minisketch` binaries `test-exhaust` and `test-exhaust-verify`. The runtimes of these are quite substantial - is there some way to do a quick sanity check as part of `make check` instead of exhaustive testing? :)

sipa commented:
@practicalswift The unit test minisketch binaries actually run forever. I need to fix that.

practicalswift commented:
@naumenkogs 

I did some robustness testing of this code by pulling in PRs #17989 (`ProcessMessage(…)` fuzzer). and #18288 (MemorySanitizer) and found an use of uninitialized memory (UUM) that is remotely triggerable.

You can reproduce the issue by pulling in the commits from those PR:s and simply run:

```
$ src/test/fuzz/process_message
…
```

The issue will be hit within a few seconds: `libFuzzer` is amazing :)

Notice also how `libFuzzer` will automatically find the newly added message names (`wtxidrelay`, `sendrecon`, `reqrecon`, `sketch`, `reqbisec` and `reconcildiff`) and probe using them all! The fuzzer harness does not need to be teached about the existence of those :)

Perhaps this UUM is the reason of the intermittent test failure you're seeing?

I encourage everybody to review (or at least Concept ACK :)) #17989 (`ProcessMessage(…)` fuzzer). and #18288 (MemorySanitizer): having them merged would be great for robustness/security.

practicalswift commented:
Needs rebase :)

naumenkogs commented:
I made some latest changes to make sure it can be plugged into a real network.
Please help with testing this by running a couple inter-connected Erlay nodes, and observing bandwidth (and ideally tx relay latency).

@practicalswift I was planning to rebase once #18044 is merged.

naumenkogs commented:
I ran 2 non-reachable nodes for around half a day, one is regular [wtxid-node](https://github.com/bitcoin/bitcoin/pull/18044) connected to legacy nodes, one is Erlay node connected to reconciliation-supporting nodes on the same host.

Bandwidth breakdown (in megabytes):

|  | Legacy (sent) | Erlay (sent) | Legacy (received) | Erlay (received) |
|----------------------------------|---------------|--------------|-------------------|------------------|
| inv | 38 | 0.4 | 22 | 5.3 |
| getdata | 6 | 4.2 | 1 | 0 |
| sketch | - | - | - | 1.2 |
| reqrecon, reqbisec, reconcildiff | - | 0.7 | - | - |
| tx, blocktxn | 3 | 0.3 | 75 | 75 |
| total (incl. other) | 48 |  7.1 | 103 | 84 |

Notice overall 40% bandwidth saving.
Please help by running similar experiments and sharing bandwidth results :)
Here's the [script](https://gist.github.com/naumenkogs/9bee7178dbe582e47d8db16f1019ddee) I hacked together for bandwidth analysis (run nodes with debug=1)

naumenkogs commented:
More experiments: now I ran the same 2 nodes for 24 hours, but connected to 16 instead of 8 nodes (yeah, all 16 peers of Erlay node support reconciliation).

Legacy wtxid-relay node spent 150 megabytes for announcements, while Erlay node spent 24 megabytes. Since these 2 days might have had different activity, it makes sense to compare the growth.. Legacy grew 2.23x (I guess due to more INVs in total), Erlay grew 1.8x. So, as expected, not only it's decent savings comparing to legacy, it also grows slower with connectivity.

Based on the following facts I also make a guess that there's no point in tweaking Erlay forward for better performance:
- only 0.0016 reconciliation failed (fully or partially) — meaning we don't underestimate much
- only 15% of the announcement bandwidth is sketches — meaning we don't overestimate much
- only 15% of the announcement bandwidth is *outbound* invs (which are much more likely to still be a little wasteful, because those peers are better connected than us and likely don't need those announemens)

Finally, I measured delay in receiving transactions. I take the earliest time I received a transaction, and take latency from there, comparing to:
- an average receiving time at all legacy nodes: 0.92s 
- at the Erlay node: 1.47s
I think this is expected and fine, as we talk in the paper, but let me know if you think differently.

mzumsande reviewed (commented):
Several existing functional tests are failing for me while synchronizing mempools via Erlay. I looked at the log of `wallet_backup.py` which has a rather simple setup of some nodes creating txes and then expecting all nodes to synchronize: 

In my run, one node would behave strangely during reconciliations, sending GETDATA for a transaction with a wtxid `000000003f946000000000003f947ae147ae147b000000000000000100000000`  even though  its peer had created a tx with a different (and actually random) wtxid. As a result, NOTFOUND is sent and the synchronization never happens. 
I haven't been able to dig in deep enough yet to understand the cause, but there still seems to be a bug somewhere.

naumenkogs commented:
@mzumsande thank you for taking a look! I'll do my best to find the cause.

Some external eyes looking at the code would definitely help to both confirm the approach and troubleshoot issues like this :)

mzumsande commented on src/net.cpp:
```diff
@@ -628,6 +628,135 @@ int CNode::GetSendVersion() const
+        if (local_tx == m_recon_state->local_short_id_mapping.end()) {
+            LogPrint(BCLog::NET, "Unknown transaction requested by short id=%d by peer=%d\n", remote_missing_short_ids[i], GetId());
+        }
+        remote_missing.push_back(local_tx->second);
```
I think that this is UB (if no wtxid for a shortid is found, there is a LogPrint, but the item is still dereferenced).
While this should explain the strange wtxids with lots of 0s, it doesn't explain the why no mapping is found to begin with.

naumenkogs replied:
I found an issue, will commit the fix later today!
Basically, sometimes reconcil. difference finding gives a false positive: it looks like a success, but some difference elements are missing.
This is just a natural property of math in minisketch. I'll submit a workaround.

naumenkogs replied:
Still not sure why it would be UB here? It just accesses the existing array element..

sipa replied:
You can't dereference local_tx when it is .end().

mzumsande replied:
I may be off, but if `local_tx == m_recon_state->local_short_id_mapping.end()`, `local_tx->second` is still accessed after the LogPrint - isn't that illegal?

naumenkogs replied:
Oh, i see you are definitely right, it's a bug!

naumenkogs commented:
Alright, all the tests should be passing now, also rebased.

There's one last test that fails consistently: p2p_segwit.py.
I'll take a look at that, and upgrade the code for recent minisketch (with reconciliation capacity checksum thing) soon.

One good idea for the future: test that legacy nodes talk OK alright to erlay nodes (I checked it only manually).

sipa commented:
@naumenkogs Please have a look at this: https://github.com/sipa/minisketch/pull/23

mzumsande commented on src/net_processing.cpp:
```diff
@@ -3417,6 +3447,274 @@ bool ProcessMessage(CNode* pfrom, const std::string& msg_type, CDataStream& vRec
+    // Record an (expected) reconciliation request with parameters to respond when time comes.
+    // All initial reconciliation responses will be done at the same time to prevent tx-related privacy leaks.
+    if (strCommand == NetMsgType::REQRECON) {
+        if (!pfrom->m_recon_state->sender) return true;
```
One more potential issue: If I understand it correctly, `m_recon_state` is null if our peer does not signal Erlay support. So we should check if `m_recon_state` is non-null before accessing its members here.

Otherwise an attacker could not signal Erlay but then send Erlay-specific messages to crash our node. Same applies for the other new messages `NetMsgType::SKETCH` etc. below.

naumenkogs replied:
Definitely an issue, will address!
I had asserts all over, and then was replacing them with returns, but this one requires 2 checks for m_recon_state and for sender :)

naumenkogs committed:
Remove node from reconciliation queue when disconnecting

naumenkogs commented:
This should be in reviewable state now: I split commits more, fixed all the tests, made it more stable, addressed feedback by @mzumsande, updated with new helpers from minisketch.

Travis failed last time because of the little issue with minisketch release build, but it should work just fine for local testing and experiments!
Will rebase again when #18044 is rebased to not create much discrepancy there.

ysangkok commented on src/protocol.h:
```diff
@@ -362,12 +402,13 @@ const uint32_t MSG_TYPE_MASK    = 0xffffffff >> 2;
     MSG_TX = 1,
     MSG_BLOCK = 2,
-    // The following can only occur in getdata. Invs always use TX or BLOCK.
+    MSG_WTX = 5, // Defined in BIP XXX
```
this could say BIP 330

ysangkok commented on src/protocol.h:
```diff
@@ -234,6 +234,46 @@ extern const char *GETBLOCKTXN;
+/**
+ * Indicates that a node prefers to relay transactions via wtxid, rather than
+ * txid.
+ * @since protocol version 70016 as described by BIP XXX.
```
also, could say BIP 330

ysangkok commented on src/minisketch/src/test-exhaust.cpp:
```diff
@@ -0,0 +1,231 @@
+        CHECK(num_roots < 1 || minisketch_decode(state, num_roots - 1, roots.data()) == -1); // Decoding with a too-low maximum should fail.
+        if (!overfill) {
+            std::sort(roots.begin(), roots.begin() + num_roots);
+//            fprintf(stderr, "Solut: ");
```
wouldn't it be better to have this guarded by an ifdef?

naumenkogs replied:
I think github.com/sipa/minisketch master branch is a better place for these comments :)

naumenkogs committed:
tests: support (inbound) connecting to mininode

naumenkogs commented:
Rebased on #18044, but we will probably get #19184 before Erlay first, so please review that first.

This PR currently doesn't pass travis tests because of the minisketch build issue after the recent patch. Should work locally and shouldn't stop people from reviewing this work :)

practicalswift commented:
It seems like the tests are failing and that a rebase is needed.

I'm super excited about Erlay! :) What is the plan ahead?

naumenkogs commented:
@practicalswift we kinda agreed that #19988 is a prerequisite, I hope Erlay becomes a priority right after :)

naumenkogs committed:
Limit transaction flooding

See how many flooding and how many reconciling
outbound peers we have, which is useful to see whether our new
peer should be using flooding or reconciliation.

It helps to save bandwidth and presumably reduces privacy leak.

naumenkogs committed:
Add a function to identify local/remote missing txs

When the sketches from both sides are combined successfully,
the diff is produced. Then this diff can (together with the local txs)
be used to identified which transactions are missing locally and remotely.

naumenkogs committed:
Add a function to announce transactions after reconciliation

Have a separate function to announce transactions to a peer (via INVs)
considering INV message limitations.

naumenkogs committed:
Handle reconciliation finalization message

Once a peer tells us reconciliation is done, we should behave as follows:
- if it was successful, just respond them with the transactions they asked
  by short ID.
- if it was a full failure, respond with all local transactions from the reconciliation
  set snapshot
- if it was a partial failure (only low or high part was failed after a bisection),
  respond with all transactions which were asked for by short id,
  and announce local txs which belong to the failed chunk.

naumenkogs committed:
Distinguish transactions to flood and to reconcile

Flooding and reconciliation are two different ways to announce
transactions to the peers.
For transactions received from inbound links, use flooding to enable
a rapid relay across the reachable nodes.

Use reconciliation (add transactions to the reconciliation sets) for
all transactions which were received from outbound links to enable
an slower but more efficient relay to unreachable nodes.

naumenkogs committed:
Announce reconciliation support

If a peer supports wtxid our node should notify them
that we are willing to participate in tx reconciliation.
At this point we generate a salt for computing
short tx IDs for reconciliations over this connection.

naumenkogs committed:
Handle reconciliation support announcement

Once we receive a message from a peer signalling
reconciliation support, we initialize the data structures
required to perform reconciliations. If we are planning to
initiate reconciliations with this peer (if it's an outbound connection),
add this peer to the reconciliation queue.

naumenkogs committed:
Reduce tx broadcast interval

Since reconciliation is naturally slower than flooding, and also
since we limit the flooding aspect, we should reduce the intervals
between broadcasts (flooding out or adding to reconciliation sets).
Otherwise, transaction relay will be too slow
(and probably less inefficient).

Note that for privacy reasons the ratio between inbound and outbound
delays matter much more than the actual delays. That ratio is preserved
here, so it is not a privacy degradation.

naumenkogs committed:
Handle extension sketch

If a peer responded to our request with a sketch extension,
attempt to decode it to find missing transactions by combining
it with the initial sketch.
If success, share/request missing transactions.
If failure, send all the transactions we had for the peer.

naumenkogs committed:
Add helper to compute sketches for tx reconciliation

Sketch is a representation of list of transaction IDs,
which enables reconciliation (an efficient sync of lists
between peers).

naumenkogs committed:
Respond to a reconciliation request

When the time comes, a node computes a sketch of the
local transactions based on the parameters sent
in the reconciliation request, and sends that sketch
to the requestor. Clear local state.

naumenkogs committed:
Add a finalize reconciliation function

This currently unused function is supposed to be used once
a reconciliation round is done. It cleans the state corresponding
to the passed reconciliation.

naumenkogs committed:
Handle reconciliation extension request

If peer failed to reconcile based on our initial response sketch,
they will ask us for a sketch extension. Store this request to respond later.

naumenkogs committed:
Prepare to responding to extension requests

Add 2 variables tracking reconciliation state:
(1) recon set snapshot and (2) capacity snapshot.
(1) is used to store transactions arrived after we sent out
an initial sketch, but before the reconciliation is over, since
these transactions should not go into a sketch extension.
(2) is used to set the capacity of the extended sketch because
it provides good estimation (efficiency/failure tradeoff).

naumenkogs commented:
Rebased, added a bunch of comments and restructured commits for better review.

Marking it as reviewable now, although my fight with Travis builds is not over yet :)

naumenkogs committed:
Respond to an extension request

Compute a sketch with extra capacity and send
those extra syndromes to the peer.
It is safe to respond to an extension request without a delay
because these requests are already limited by initial requests.

naumenkogs commented:
Me and @sipa decided to switch to "extension sketches": instead of requesting a bisection, someone who failed to decode the difference from the initial sketch, may ask for an extension of the initial sketch.
A responder then have to send syndromes of higher order only.

The main motivation is it seems like bisection adds a lot of complexity (see b5c92a41e4cc0599504cf838d20212f1a403e573).
Extensions have all the same properties after all, but they're more CPU-costly. We think it's fine for sketches of small capacity (what we expect to have here) to ignore this disadvantage.

So, for now, please review up to a30e861c9cada8fbc400d07255516839480d9e9d (but also feel free to review further if you're curious).

I will replace bisection with extensions and update the BIP soon.

naumenkogs committed:
Handle reconciliation sketch and successful decoding

Send/request missing transactions, clear the state,
and send a finalization message.

naumenkogs committed:
Request extension if decoding failed

If after decoding a reconciliation sketch it turned out
to be insufficient to find set difference, request extension.
Store the initial sketches so that we are able to process
extension sketch.

naumenkogs commented:
Okay, now the [BIP PR](https://github.com/bitcoin/bitcoin/pull/18261#issuecomment-600388543) is up-to-date with this code, and this stuff is ready for full review again.

ariard commented:
Starting the review by digging into the BIP :)

Few remarks :
1) "improves privacy as a side-effect", it would be great to explain and document this claim, maybe add a section in Rationale ?
2) "truncated transaction IDs",  AFAIU, Erlay introduces new shortened transaction identifiers for regular tx-relay, maybe it could be better to call them "truncated WTXID" ? If BIP 330 implies BIP 339, it could be better to remove references to transaction ID or at least when it's ambiguous
3) You mentioned switching to extension sketch and mentioned the updated BIP was ready for review, I think you need to update the "intented protocol flow" schema, remove `reqbisec` and update Rationale ?
4) You might swap the "Local State" section before messages, it could be easier to explain q-coefficient before using them in New messages

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -2370,6 +2370,28 @@ static void ProcessGetCFCheckPt(CNode& peer, CDataStream& vRecv, const CChainPar
+ * No need to add transactions to peer's filter or do checks
+ * because it was already done when adding to the reconciliation set.
+ */
+void static AnnounceTxs(std::vector<uint256> remote_missing_wtxids, CNode* pto, CNetMsgMaker msgMaker, CConnman* connman)
```
I assume you could pass this vector by (const) reference?  (And maybe all the other stuff can be passed by reference too...).

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -2370,6 +2370,28 @@ static void ProcessGetCFCheckPt(CNode& peer, CDataStream& vRecv, const CChainPar
+{
+    if (remote_missing_wtxids.size() != 0) {
+        std::vector<CInv> remote_missing_invs;
+        for (uint256 wtxid: remote_missing_wtxids) {
```
nit: I think this could be a const reference as well, and probably there's some modern c++ you could write so that we're not making extra copies in this loop, someone else can chime in and explain how to do that...

sdaftuar reviewed (commented):
Concept ACK on Erlay -- thanks for working on this.

My review approach is to focus on the non-minisketch portions of this PR for now.  I've read through the BIP draft and gone through the first 5 commits.  As this is a big PR I think it'll take me a while to go through the rest, so I'm leaving these comments for now and will come back to this later when I have time for additional review.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -827,6 +833,30 @@ static void FindNextBlocksToDownload(NodeId nodeid, unsigned int count, std::vec
+        if (!pnode->m_tx_relay)
+            return;
+        if (!pnode->IsFullOutboundConn() && !pnode->IsManualConn())
+            return;
```
Haha, I was planning to bring this up later in the review cycle. I didn't expect someone to find it that fast :)
Yeah, what you describing is possible.
Perhaps we should have a loop checking that we have enough (8) flood outbound peers, and then if not, switch some existing recon peers to flooding.

Will add a commit for this.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -827,6 +833,30 @@ static void FindNextBlocksToDownload(NodeId nodeid, unsigned int count, std::vec
+{
+    size_t result = 0;
+    m_connman.ForEachNode([flooding, &result](CNode* pnode) {
+        if (!pnode->m_tx_relay)
```
Fixed this in the first batch of commits, will also fix for the rest.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -2559,6 +2617,62 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
+        // Reconcile with all outbound peers supporting reconciliation (even if we flood to them),
+        // to not miss transactions they have for us but won't flood.
+        if (peer->m_recon_state->m_responder) {
+            m_recon_queue.push_back(&pfrom);
```
Probably gonna be restrictive for now.

naumenkogs replied:
Updating the code and the BIP

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -899,6 +926,14 @@ void PeerManager::UpdateNextReconRequest(std::chrono::microseconds now)
+std::chrono::microseconds PeerManager::NextReconRespond(std::chrono::microseconds now)
+{
+    if (m_next_recon_respond < now) {
+        m_next_recon_respond = now + RECON_RESPONSE_INTERVAL;
```
I'm wondering if this might be better as a poisson-timer, but I don't quite understand the implications of all this yet to reach a conclusion.  Intuition: with a fixed delay like this, an adversary with 2 inbound connections can use set reconciliation to measure the set difference between the node's transactions-to-announce in known and arbitrarily precise time slices.

naumenkogs replied:
I also don't have a good answer, but it's not that simple of an attack: we [have](https://github.com/bitcoin/bitcoin/pull/18261/files#r541017754) Poisson on adding to reconciliation sets.

Having another poison here, well, I don't think it makes anything worse. Maybe you're right.

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -610,6 +625,42 @@ struct Peer {
+            for (uint256 wtxid: local_set) {
+                uint32_t short_txid = ComputeShortID(wtxid);
+                short_ids.push_back(short_txid);
+                m_local_short_id_mapping.insert(std::pair<uint32_t, uint256>(short_txid, wtxid));
```
Should the `m_local_short_id_mapping` be cleared out in this function?

naumenkogs replied:
Why so early? The point of it is to refer to it later, when we are requested something by short ID.
That's why it should be kept until `FinalizeReconciliation`. 

Although as currently implemented, I notice there's a bug, because we may lose short IDs of the transactions received during extension when calling this after the extension is over. I need to fix this part.

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -4200,6 +4209,53 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
+        vRecv >> skdata;
+
+        // Attempt to decode the received sketch with a local sketch.
+        if (skdata.size() / BYTES_PER_SKETCH_CAPACITY > MAX_SKETCH_CAPACITY) return;
```
I think this is a protocol violation and leaves the link in a permanently broken reconciliation state, since a reconcildiff would never be sent. Perhaps we should disconnect the peer if this happens?

naumenkogs replied:
Not really permanently broken, they still have a chance to send us a proper sketch. But yeah, either finalizing or disconnecting would work. Making it disconnect for now.

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -4200,6 +4209,53 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
+        minisketch_deserialize(remote_sketch, remote_sketch_serialized);
+
+        minisketch* working_sketch; // Contains sketch of the set difference
+        minisketch* local_sketch = peer->m_recon_state->ComputeSketch(peer->m_recon_state->m_local_set, remote_sketch_capacity);
```
I think we need to call `minisketch_destroy` on these somewhere to avoid a memory leak right?

sdaftuar commented on src/net_processing.cpp:
```diff
@@ -4252,6 +4262,16 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
+            // Initial reconciliation failed.
+            // Store the received sketch and the local sketch, request extension.
+
+            assert(remote_sketch != nullptr);
```
Is it possible for the peer to send us a message that causes `remote_sketch` to be nullptr here?

naumenkogs replied:
Adding protection for this case.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -2087,7 +2105,7 @@ void PeerManager::ProcessOrphanTx(std::set<uint256>& orphan_work_set)
         if (AcceptToMemoryPool(m_mempool, state, porphanTx, &removed_txn, false /* bypass_limits */)) {
             LogPrint(BCLog::MEMPOOL, "   accepted orphan tx %s\n", orphanHash.ToString());
-            RelayTransaction(orphanHash, porphanTx->GetWitnessHash(), m_connman);
+            RelayTransaction(orphanHash, porphanTx->GetWitnessHash(), m_connman, true);
```
I think you're right, although we should be extra careful to make sure no privacy leak is introduced w.r.t topology.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -123,12 +123,12 @@ static constexpr std::chrono::hours AVG_LOCAL_ADDRESS_BROADCAST_INTERVAL{24};
 /** Average delay between trickled inventory transmissions in seconds.
  *  Blocks and peers with noban permission bypass this, outbound peers get half this delay. */
-static const unsigned int INVENTORY_BROADCAST_INTERVAL = 5;
+static const unsigned int INVENTORY_BROADCAST_INTERVAL = 2;
```
Yeah, so I think the ratio matters most. It was 2.5 and now it's 2, so a bit different.
But I'm also not 100% if this difference matters, while preserving 2.5 would require operating with floats now, so I decided to not, at least yet.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -2562,6 +2620,74 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
+            // or want to be reconciliation senders (request our sketches).
+            // Just ignore SENDRECON and use normal flooding for transaction relay with them.
+            if (recon_sender) return;
+            if (!recon_responder) return;
```
Right, this implementation prioritized simplicity here.
Perhaps, I should add sending RECONACK, and also think whether at least some of the non-matching scenarios are compatible.

naumenkogs replied:
@sipa good idea, ack.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -123,12 +123,12 @@ static constexpr std::chrono::hours AVG_LOCAL_ADDRESS_BROADCAST_INTERVAL{24};
 static constexpr unsigned int INVENTORY_BROADCAST_PER_SECOND = 7;
 /** Maximum number of inventory items to send per transmission. */
-static constexpr unsigned int INVENTORY_BROADCAST_MAX = INVENTORY_BROADCAST_PER_SECOND * INVENTORY_BROADCAST_INTERVAL;
+static constexpr unsigned int INVENTORY_BROADCAST_MAX = INVENTORY_BROADCAST_PER_SECOND * INVENTORY_BROADCAST_INTERVAL * 4;
```
`INVENTORY_BROADCAST_PER_SECOND` is probably derived from the block size, so it shouldn't be changed.
`INVENTORY_BROADCAST_INTERVAL` is just how long we wait between reconciliations with different peers (say 2 seconds).

Now, imagine, a non-reachable node (which only learns from reconciliations), reconciled with a dysfunctional peer. So the stack of what it should learn during the next successful reconciliation is twice the regular size.
And `INVENTORY_BROADCAST_MAX` should better accommodate for these cases, because otherwise what we'll get is at least inefficiency.

"*4" is just an intuitive amortization for these cases.

I probably should add the explanation to a commit message once we agree this makes sense.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -3993,6 +4028,22 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
+        peer->m_recon_state->m_incoming_recon = Peer::ReconState::ReconPhase::INIT_REQUESTED;
+        peer->m_recon_state->m_remote_set_size = peer_recon_set_size;
+        peer->m_recon_state->m_remote_q = double(peer_q / Q_PRECISION);
+        peer->m_recon_state->m_next_recon_respond = NextReconRespond(current_time);
```
I think there's no real difference with this design though?
We gain privacy by not allowing 2 peers to monitor our mempool, because we respond to them at the same time (every 2 seconds).

Snapshoting earlier would hide the transactions received during the last 2 seconds and disable this kind of monitoring, but then reconciliations would not be so "fresh" (txs received during last 2s go into the next one).
Not sure this kind of monitoring is that dangerous, especially since we *also* have a Poisson on adding to recon sets.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -182,6 +183,20 @@ static constexpr std::chrono::microseconds RECON_REQUEST_INTERVAL{std::chrono::s
+*/
+static constexpr unsigned int RECON_FALSE_POSITIVE_COEF = 16;
+static_assert(RECON_FALSE_POSITIVE_COEF <= 256,
+    "Reducing reconciliation false positives beyond 1 in 2**256 is not supported");
```
Yeah, definitely, I have the same feel. Perhaps it better belongs to ReconState{}.

sipa commented on src/net_processing.cpp:
```diff
@@ -146,6 +147,65 @@ static constexpr uint32_t MAX_GETCFILTERS_SIZE = 1000;
+/** Static component of the salt used to compute short txids for transaction reconciliation. */
+static const std::string RECON_STATIC_SALT = "Tx Relay Salting";
+/** Used to convert a floating point reconciliation coefficient q to an int for transmission. */
+static constexpr double Q_PRECISION{2 << 12};
```
BIP comment: why 12 bits? You could have 16 bits of accuracy with the 2 bytes that it takes.

naumenkogs replied:
q is computed as `(total_missing - abs(local_missing - remote_missing)) / min_size`

So the bounds on q are [0...2] I think?
if abs = 0, total_missing = 2 * min_size is an upper bound.
If abs !=0, well. Ideally, we need to solve this equation properly :)

So it should be (2 << 14) - 1 I think assuming [0..2] for now.

sipa commented on src/net_processing.cpp:
```diff
@@ -146,6 +147,65 @@ static constexpr uint32_t MAX_GETCFILTERS_SIZE = 1000;
+static constexpr std::chrono::microseconds RECON_RESPONSE_INTERVAL{std::chrono::seconds{2}};
+/** The size of the field, used to compute sketches to reconcile transactions (see BIP-330). */
+static constexpr unsigned int RECON_FIELD_SIZE = 32;
+static_assert(RECON_FIELD_SIZE % 8 == 0, "Field size should be divisible by 8");
```
Why is this strictly needed? (apart from the BIP specifying that it is 32, so any other value is protocol breaking).

sipa commented on src/net_processing.cpp:
```diff
@@ -451,7 +511,305 @@ struct Peer {
+            return sketch;
+        }
+
+        minisketch* ComputeExtendedSketch() {
```
I've PR'ed https://github.com/sipa/minisketch/pull/28 that adds a safer C++ wrapper around the minisketch* type. You may want to use it instead of doing manual memory management for them. It also adds some convenience functions for dealing with fpbits (CreateFP and DecodeFP).

sipa commented on test/functional/p2p_erlay.py:
```diff
@@ -0,0 +1,596 @@
+    """Compute 2*x in GF(2^FIELD_BITS)"""
+    return (x << 1) ^ (FIELD_MODULUS if x.bit_length() >= FIELD_BITS else 0)
+
+def mul(x, y):
```
I've PR'ed https://github.com/sipa/minisketch/pull/26 which adds a pure Python (slow) full reimplementation of minisketch. You may want to use that instead (it also supports decoding).

sipa reviewed (commented):
First set of comments.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -461,6 +465,63 @@ struct Peer {
+        double m_local_q;
+    };
+
+    RecursiveMutex cs_recon_state;
```
I'm not 100% sure how to do it right, but simple Mutex doesn't seem to work when `GetFloodingOutboundsCount` is called in handling `NetMsgType::SENDRECON`. It just hangs.

jnewbery commented on src/net.h:
```diff
@@ -556,9 +556,11 @@ class CNode
-        // Set of transaction ids we still have to announce.
-        // They are sorted by the mempool before relay, so the order is not important.
-        std::set<uint256> setInventoryTxToSend;
+        // Set of transaction ids we still have to announce, and whether we may flood them:
```
This is no longer a set.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -146,6 +153,37 @@ static constexpr uint32_t MAX_GETCFILTERS_SIZE = 1000;
+ * reconciliation metadata (sketch sizes etc.), which would nullify the efficiency.
+ * Less frequent reconciliations would introduce high transaction relay latency.
+ */
+static constexpr std::chrono::microseconds RECON_REQUEST_INTERVAL{std::chrono::seconds{16}};
```
```suggestion
static constexpr std::chrono::microseconds RECON_REQUEST_INTERVAL{16s};
```

Saves repeating the chrono type

jnewbery commented on src/net_processing.cpp:
```diff
@@ -146,6 +153,37 @@ static constexpr uint32_t MAX_GETCFILTERS_SIZE = 1000;
+static constexpr std::chrono::microseconds RECON_RESPONSE_INTERVAL{std::chrono::seconds{2}};
+
+/** Represents a reconciliation result, used to decide what to do when the reconciliation is over. */
+enum ReconResult {
```
Any reason not to use a bool here?

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2267,6 +2355,29 @@ static void ProcessGetCFCheckPt(CNode& peer, CDataStream& vRecv, const CChainPar
+ * No need to add transactions to peer's filter or do checks
+ * because it was already done when adding to the reconciliation set.
+ */
+void static AnnounceTxs(const std::vector<uint256>& remote_missing_wtxids, CNode& pto, const CNetMsgMaker& msgMaker, CConnman& connman)
```
It's easy enough to instantiate a new `CNetMsgMaker` from the CNode, rather than passing it as an argument. If you make this function a member of `PeerManagerImpl` you can also avoid passing the `CConnman&`.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2267,6 +2355,29 @@ static void ProcessGetCFCheckPt(CNode& peer, CDataStream& vRecv, const CChainPar
+ */
+void static AnnounceTxs(const std::vector<uint256>& remote_missing_wtxids, CNode& pto, const CNetMsgMaker& msgMaker, CConnman& connman)
+{
+    if (remote_missing_wtxids.size() != 0) {
```
Consider inverting this if condition to make this a guard clause and avoid deep indentations below.

```suggestion
    if (remote_missing_wtxids.size() == 0) return;
```

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2267,6 +2355,29 @@ static void ProcessGetCFCheckPt(CNode& peer, CDataStream& vRecv, const CChainPar
+void static AnnounceTxs(const std::vector<uint256>& remote_missing_wtxids, CNode& pto, const CNetMsgMaker& msgMaker, CConnman& connman)
+{
+    if (remote_missing_wtxids.size() != 0) {
+        std::vector<CInv> remote_missing_invs;
```
Perhaps reserve max(size of remote_missing_wtxids, MAX_INV_SIZE) to avoid reallocations.

naumenkogs replied:
I think this should be min, because the vector never exceeds MAX_INV_SIZE.

jnewbery replied:
yes, you're right. Should be min.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -3751,6 +3943,165 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
+
+    // Record an (expected) reconciliation request with parameters to respond when time comes.
+    // All initial reconciliation responses will be done at the same time to prevent tx-related privacy leaks.
+    if (msg_type == NetMsgType::REQRECON) {
```
A bit of space/comments in this code block would make it easier to read.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -3751,6 +3943,165 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
+
+    // Among transactions requested by short ID here, we should send only those transactions
+    // sketched (stored in local set snapshot), because otherwise we would leak privacy (mempool content).
+    if (msg_type == NetMsgType::RECONCILDIFF) {
```
Again, spacing/comments would make this more legible.

jnewbery commented on src/net_processing.h:
```diff
@@ -11,6 +11,7 @@
 #include <sync.h>
 #include <txrequest.h>
 #include <validationinterface.h>
+#include <minisketch/include/minisketch.h>
```
sort

jnewbery commented on src/net_processing.h:
```diff
@@ -31,6 +32,19 @@ static const bool DEFAULT_PEERBLOOMFILTERS = false;
 static const int DISCOURAGEMENT_THRESHOLD{100};
+/** The size of the field, used to compute sketches to reconcile transactions (see BIP-330). */
+static constexpr unsigned int RECON_FIELD_SIZE = 32;
+static constexpr unsigned int BYTES_PER_SKETCH_CAPACITY = RECON_FIELD_SIZE / 8;
```
This doesn't need to be in the header, since it's only used inside net_processing.cpp.

jnewbery commented on src/net_processing.h:
```diff
@@ -89,7 +103,314 @@ struct Peer {
+         * A sketch has a capacity meaning it allows reconciling at most a certain number of elements. (see BIP-330).
+         * Considering whether we are going to send a sketch to a peer or use locally, we estimate the set difference.
+         */
+        Minisketch ComputeSketch(const std::set<uint256> local_set, uint16_t& capacity)
```
Inside a structure inside `Peer` feels like the wrong place for a lot of this complex logic. I think ideally, `Peer` would continue to be a struct (i.e. data members only) and the logic would be contained in a separate module.

jnewbery commented on src/net_processing.h:
```diff
@@ -89,7 +103,314 @@ struct Peer {
+        }
+    };
+
+    RecursiveMutex cs_recon_state;
```
Don't use old cs nomenclature for mutexes:

```suggestion
    RecursiveMutex m_recon_state_mutex;
```

Also prefer to use a `Mutex` over a `RecursiveMutex` in new code.

naumenkogs replied:
I think I can't make it non-recursive because of the call inside `GetFloodingOutboundsCount`. How would you suggest to overcome this issue?

jnewbery commented on src/net_processing.h:
```diff
@@ -232,6 +553,38 @@ class PeerManager final : public CValidationInterface, public NetEventsInterface
+     * is predictable. This queue is used to maintain the order of
+     * peers chosen for reconciliation.
+     */
+    RecursiveMutex cs_recon_queue;
```
Again, prefer a non-reenrant mutex and don't use the cs nomenclature.

naumenkogs replied:
I can switch to non-recursive if UpdateNextReconRequest() takes `m_recon_queue.size()` as an argument, instead of accessing it inside. Do you think that would be preferable?

jnewbery replied:
I haven't looked into your branch in great detail, but it looks like `UpdateNextReconRequest()` is only called in one place, which already holds `m_recon_queue_mutex`. Could you make the function `EXCLUSIVE_LOCKS_REQUIRED(m_recon_queue_mutex)` and not retake the mutex inside?

jnewbery commented on test/functional/p2p_erlay.py:
```diff
@@ -0,0 +1,595 @@
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test reconciliation-based transaction relay protocol.
+
```
Why this space?

jnewbery commented on test/functional/p2p_erlay.py:
```diff
@@ -0,0 +1,595 @@
+
+"""
+
+from test_framework.p2p import P2PDataStore, p2p_lock
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal, hex_str_to_bytes
+from test_framework.messages import (
+    msg_inv, msg_getdata,
+    msg_sendrecon, msg_reqrecon,
+    msg_sketch, msg_reqsketchext, msg_reconcildiff,
+    CTransaction, CInv
+)
+from test_framework.siphash import siphash256
+from test_framework.key import TaggedHash
+from enum import IntEnum
+from io import BytesIO
+import random
+import hashlib
+import time
+import struct
```
Please sort, with standard library imports first.

jnewbery commented on test/functional/p2p_erlay.py:
```diff
@@ -0,0 +1,595 @@
+        msg = msg_getdata(inv=[CInv(MSG_WTX, h=wtxid) for wtxid in ask_wtxids])
+        self.send_message(msg)
+
+class ReconResult(IntEnum):
+    RECON_FAILED = 0
+    RECON_SUCCESS = 1
```
Why not a bool?

jnewbery commented on test/functional/test_framework/messages.py:
```diff
@@ -121,6 +121,29 @@ def ser_uint256(u):
     return rs
 
 
+def deser_uint128(f):
+    r = 0
+    for i in range(4):
+        t = struct.unpack("<I", f.read(4))[0]
+        r += t << (i * 32)
+    return r
+
+
+def ser_uint128(u):
+    rs = b""
+    for i in range(4):
+        rs += struct.pack("<I", u & 0xFFFFFFFF)
+        u >>= 32
+    return rs
```
I think that these are unneeded now that truncated txids aren't used

jnewbery commented on test/functional/test_framework/messages.py:
```diff
@@ -178,6 +201,49 @@ def ser_uint256_vector(l):
+        t = struct.unpack("<B", f.read(1))[0]
+        r.append(t)
+    return r
+
+
```
why two blank lines?

jnewbery commented on test/functional/test_framework/p2p.py:
```diff
@@ -27,6 +27,7 @@
 import struct
 import sys
 import threading
+import socket
```
sort

jnewbery commented on test/functional/test_framework/p2p.py:
```diff
@@ -59,6 +60,11 @@
     msg_pong,
     msg_sendaddrv2,
     msg_sendcmpct,
+    msg_sendrecon,
```
sort

jnewbery commented on test/functional/test_framework/p2p.py:
```diff
@@ -107,6 +113,11 @@
     b"verack": msg_verack,
     b"version": msg_version,
     b"wtxidrelay": msg_wtxidrelay,
+    b"sendrecon": msg_sendrecon,
```
sort

jnewbery commented on test/functional/test_framework/p2p.py:
```diff
@@ -379,6 +408,7 @@ def on_sendaddrv2(self, message): pass
     def on_sendcmpct(self, message): pass
     def on_sendheaders(self, message): pass
     def on_tx(self, message): pass
+    def on_sendrecon(self, message): pass
```
sort

jnewbery reviewed (commented):
Just some style comments so far. A couple of high-level points:

1. I agree with the comment here https://github.com/bitcoin/bitcoin/pull/18261#discussion_r541049259 that we should aim to separate this from the rest of net_processing. This PR currently has about 850 LOC change in net_processing.{cpp|h}, out of a total of around 5700 lines. That means ~15% of net_processing is erlay code after this PR (and over half of net_processing.h is erlay specific). Would it be possible to split the erlay logic into its own subcomponent? One immediate benefit for you is that there would be far fewer disruptive rebases if you did that.
2. Could the minisketch code and tests be split into their own PR? It seems like there is pretty wide support for incorporating erlay. Reviewing and merging minisketch first would make this PR a lot smaller and more focused on just the p2p changes.

naumenkogs committed:
Prepare to distinguish transactions to flood and to reconcile

We need assign and keep track of the way we're going to relay a transaction
based on the connection type the transaction arrived from, to
enable transaction reconciliation protocol.

naumenkogs committed:
Add a function to get wtxids by shortids

At the end of a reconciliation round, a peer may ask us
for transactions by their short id. Add a function for a local
lookup short_id->wtxid.

naumenkogs committed:
Request tx reconciliation as scheduled

If the peer is next in the queue for reconciliation,
and enough time is passed from the previous
reconciliation, consider a peer for reconciliation.

If there is no pending reconciliation request to the peer,
send one, otherwise just move to the end of queue and
update global next reconciliation request time.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -219,7 +232,20 @@ struct Peer {
+    const uint64_t m_local_recon_salt;
+
+    Mutex m_recon_state_mutex;
+    /// nullptr if we're not reconciling (neither passively nor actively) with this peer.
```
Please use doxygen style comments.

jnewbery commented on src/net_processing.h:
```diff
@@ -7,6 +7,7 @@
 #define BITCOIN_NET_PROCESSING_H
 
 #include <net.h>
+#include <reconciliation.h>
```
This isn't needed in the header.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -3192,7 +3402,13 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+            // Flood those transactions which were received either via flooding, or inbound reconciliation,
+            // but NOT via outbound reconciliation. Flooding then is mainly used for initial propagation
+            // of new transactions across a network of reachable nodes quickly.
+            LOCK(peer->m_recon_state_mutex);
+            bool flood = !(peer->m_recon_state && peer->m_recon_state->m_responder);
```
Use a `WITH_LOCK()` macro or code block here to avoid holding the mutex for the rest of this block.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -3944,6 +4160,163 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         return;
     }
 
+    std::chrono::microseconds current_time = GetTime<std::chrono::microseconds>();
```
Please place this inside the `REQRECON` message handling code block (we probably want to split each message handler into its own function in the future, so we should minimize the use of code outside those code blocks.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -4543,6 +4916,7 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
         // Message: inventory
         //
         std::vector<CInv> vInv;
+        PeerRef peer = GetPeerRef(pto->GetId());
```
Not required. There is already a local `peer` in this function.

jnewbery commented on src/net.h:
```diff
@@ -688,12 +690,12 @@ class CNode
     }
 
-    void PushTxInventory(const uint256& hash)
+    void PushTxInventory(const uint256& hash, const bool flood)
```
No need to mark pass-by-value arguments as const.

jnewbery commented on src/net_processing.h:
```diff
@@ -69,6 +70,6 @@ class PeerManager : public CValidationInterface, public NetEventsInterface
 
 /** Relay transaction to every node */
-void RelayTransaction(const uint256& txid, const uint256& wtxid, const CConnman& connman) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+void RelayTransaction(const uint256& txid, const uint256& wtxid, const CConnman& connman, bool flood) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
Consider making `flood=false` the default, so that call sites don't need to be updated for this change.

jnewbery commented on src/reconciliation.h:
```diff
@@ -0,0 +1,335 @@
+#include <minisketch/include/minisketch.h>
```
Add copyright boilerplate

jnewbery replied:
This file should include all the standard library headers that it uses (eg chrono, set, etc)

jnewbery replied:
Add include guards

jnewbery commented on src/reconciliation.h:
```diff
@@ -0,0 +1,335 @@
+#include <minisketch/include/minisketch.h>
+
+/** Static component of the salt used to compute short txids for transaction reconciliation. */
```
Can you try to limit these lines to 80 or 100 chars to avoid wrapping.

naumenkogs replied:
This particular line is less than 100 chars, so I assume you referred to the other ones which are longer than 100. I see a couple, gonna limit them.

jnewbery commented on src/reconciliation.h:
```diff
@@ -0,0 +1,335 @@
+    ReconState(bool requestor, bool responder, bool flood_to, uint64_t k0, uint64_t k1) :
+        m_requestor(requestor), m_responder(responder), m_flood_to(flood_to), m_k0(k0), m_k1(k1), m_local_q(DEFAULT_RECON_Q) {}
+
+    /// Whether this peer will send reconciliation requests.
```
doxygen

jnewbery commented on src/reconciliation.h:
```diff
@@ -0,0 +1,335 @@
+ * sequence, nodes are supposed to exchange transactions, so that both of them
+ * have all relevant transactions. For more protocol details, refer to BIP-0330.
+ */
+struct ReconState {
```
This should really be a class since it has function members.

mzumsande commented on src/net_processing.cpp:
```diff
@@ -3358,7 +3375,13 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+            // but NOT via outbound reconciliation. Flooding then is mainly used for initial propagation
+            // of new transactions across a network of reachable nodes quickly.
+            LOCK(peer->m_recon_state_mutex);
+            bool flood = !(peer->m_recon_state && peer->m_recon_state->m_responder);
```
Does this work as intended in the intermediate stage where not all of our peers support Erlay? If we are a non-reachable node that supports Erlay but has just one older pre-Erlay outbound peer - wouldn't we flood all of its transactions also to our Erlay-supporting peers because `flood == true` here? (Considering the goal that "only well-connected publicly reachable nodes flood transactions to other publicly reachable nodes via outbound connections").

naumenkogs replied:
"only well-connected publicly reachable nodes flood transactions to other publicly reachable nodes via outbound connections" only is an end-goal when most of the network implement Erlay.

In the meanwhile, I think this behavior is better because it's most intuitive: for legacy peers, we support legacy behavior (flood).
Technically, changing this condition wouldn't break stuff, but I think it's just less intuitive.

I probably should expand the comment to address this misunderstanding though.

mzumsande replied:
But the legacy behavior also affects Erlay peers (because if a non-reachable node received transactions from legacy  peers, it would act as if it was a reachable hub and forward them via flooding to other outbound peers, including Erlay ones).

Apart from this specific spot that could easily be adjusted if needed, I think that when there are two relay mechanisms with different time scales existing in parallel, this could lead to some non-intuitive consequences - such as legacy relay being dominant and cannibalizing Erlay relay way into the regime when the majority of nodes already support Erlay.
 
Would it maybe be worth doing a mixed-network simulation run and measuring the respective contributions at different ratios of Erlay to legacy nodes, if this is possible within the existing simulation framework?

naumenkogs replied:
@mzumsande sorry I just saw this question. Yeah, that might be useful, once I'm done with big refactor for this...
Intuitively, I think this shouldn't be the case legacy peers cannibalize efficiency I think.

mzumsande commented on src/protocol.h:
```diff
@@ -267,6 +267,13 @@ extern const char* WTXIDRELAY;
+ * for a peer to construct a set sketch.
+ * Peer should respond with "sketch" message.
+ */
+extern const char *REQRECON;
```
This message is called "reqrecon**cil**" in the BIP.

glozow commented on src/net_processing.cpp:
```diff
@@ -3944,6 +4160,163 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+        LOCK(peer->m_recon_state_mutex);
+        if (peer->m_recon_state == nullptr) return;
+        if (!peer->m_recon_state->m_requestor) return;
+        if (peer->m_recon_state->m_incoming_recon != RECON_NONE) return;
```
Question: is it possible for `peer->m_recon_state->m_outgoing_recon != RECON_NONE`, i.e. we sent them a `REQRECON` around the same time they sent us one?

ariard reviewed (commented):
Still doing a first parse of the PR, some comments are really minors but I think few more substantial.

At first sight, I think this PR could benefit from better code organization (e.g split requester/responder state in different struct, encapsulate processing of new messages, ...) but need to finish the high-level review before to propose.

dergoegge reviewed (commented):
Concept ACK on Erlay - Reducing bandwidth usage and increasing connectivity is great! The slightly increased transaction propagation times are an OK tradeoff IMO.

unseddd commented on src/net.h:
```diff
@@ -551,9 +551,11 @@ class CNode
+        // - true: flood unless peer correctly negotiated reconciliation and we didn't choose that peer for flooding.
+        // - false: use reconciliation unless it's not supported by the peer, flood otherwise.
+        // Transactions are sorted by the mempool before relay, so the order is not important.
+        std::map<uint256, bool> m_transactions_to_announce;
```
nit: since order is not important, maybe std::unordered_map for efficiency?

unseddd commented on src/reconciliation.h:
```diff
@@ -0,0 +1,335 @@
+ * reconciliation metadata (sketch sizes etc.), which would nullify the efficiency.
+ * Less frequent reconciliations would introduce high transaction relay latency.
+ */
+static constexpr std::chrono::microseconds RECON_REQUEST_INTERVAL{16s};
```
nit: use smth like:

```
static constexpr std::chrono::microseconds RECON_REQUEST_INTERVAL{16000000};
```

using the `"" s` operator requires a `using namespace std::chrono_literals` or similar: [https://en.cppreference.com/w/cpp/chrono/operator%22%22s](https://en.cppreference.com/w/cpp/chrono/operator%22%22s).

Same for `RECON_RESPONSE_INTERVAL`.

unseddd commented:
Added a couple fuzz harnesses in my [erlay](https://github.com/unseddd/bitcoin/tree/erlay) branch, if you want them.

Also contains some other minor fixups, like a change for [src/test/fuzz/net.cpp](https://github.com/bitcoin/bitcoin/blob/master/src/test/fuzz/net.cpp#L89) (for the `flood` param):

```diff
diff --git a/src/test/fuzz/net.cpp b/src/test/fuzz/net.cpp
index 31b99600e..7d48a7ec3 100644
--- a/src/test/fuzz/net.cpp
+++ b/src/test/fuzz/net.cpp
@@ -86,7 +86,8 @@ FUZZ_TARGET_INIT(net, initialize_net)
                 node.AddKnownTx(inv_opt->hash);
             },
             [&] {
-                node.PushTxInventory(ConsumeUInt256(fuzzed_data_provider));
+                node.PushTxInventory(ConsumeUInt256(fuzzed_data_provider), false);
+                node.PushTxInventory(ConsumeUInt256(fuzzed_data_provider), true);
             },
             [&] {
                 const std::optional<CService> service_opt = ConsumeDeserializable<CService>(fuzzed_data_provider);
```

unseddd commented:
I understand that the requestor being the one to compute the sketches protects public nodes against potential DoS. Are there any protections for requestor(s)?

Thinking about the scenario where one public node requests a sketch from a malicious public node who returns high diff sketches as a DoS.

jnewbery commented:
I think the approach needs rethinking for the reasons stated here: https://github.com/bitcoin/bitcoin/pull/18261#pullrequestreview-570100861. The Erlay logic in this PR is spread across many functions in net_processing.cpp and reconcilliation.h, which means that it's impossible to test that logic in isolation. I suppose that's why there aren't any unit tests included in this PR.

I think it would be much cleaner to have a fully separate module, similar to how txrequest was separated from the rest of net_processing. That module maintains its own private state and exposes a limited number of interface methods to net_processing. That means that it's possible to very thoroughly unit test and fuzz test that module in isolation.

jonatack commented:
Needs rebase.

jonatack commented on src/minisketch/include/minisketch.h:
```diff
@@ -0,0 +1,306 @@
+        ssize_t ret = minisketch_decode(m_minisketch.get(), max_elements, result.data());
+        if (ret == -1) return {};
+        result.resize(ret);
+        return std::move(result);
```
```
In file included from ./reconciliation.h:1,
                 from ./net_processing.h:10,
                 from rpc/net.cpp:13:
./minisketch/include/minisketch.h: In member function ‘std::optional<std::vector<long unsigned int> > Minisketch::Decode(size_t) const’:
./minisketch/include/minisketch.h:293:25: warning: redundant move in return statement [-Wredundant-move]
  293 |         return std::move(result);
      |                ~~~~~~~~~^~~~~~~~
./minisketch/include/minisketch.h:293:25: note: remove ‘std::move’ call
```

naumenkogs commented on src/net.h:
```diff
@@ -551,9 +551,11 @@ class CNode
-        std::set<uint256> setInventoryTxToSend;
+        // Transaction ids we still have to announce, and whether we may flood them:
+        // - true: flood unless peer correctly negotiated reconciliation and we didn't choose that peer for flooding.
+        // - false: use reconciliation unless it's not supported by the peer, flood otherwise.
```
How to reflect that if it's "false, use reconciliation", it still can be flooded, in case if reconciliation is not supported for a given peer?

ariard replied:
See this [comment](https://github.com/bitcoin/bitcoin/pull/18261#discussion_r573107239) which I believe is superseding this discussion.

jonatack commented on src/net_processing.cpp:
```diff
@@ -4559,22 +4933,23 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
 
             if (pto->m_tx_relay != nullptr) {
                 LOCK(pto->m_tx_relay->cs_tx_inventory);
+                LOCK(peer->m_recon_state_mutex);
```
After building with both gcc 10.2.1 and clang 9, functional tests are failing with `AssertionError: Unexpected stderr Assertion failed: detected inconsistent lock order for 'peer->m_recon_state_mutex' in net_processing.cpp:4936 (in thread 'msghand')`

ariard commented on src/net_processing.cpp:
```diff
@@ -219,7 +219,16 @@ struct Peer {
+     * Also, salts should be different per-connection to prevent halting
+     * of relay of particular transactions due to collisions in short IDs.
+     */
+    const uint64_t m_local_recon_salt;
```
6103b5cd

I think you can make this part of `ReconState`, at least a tiny memory saved for non-reconciliation peers.

pinheadmz commented:
I had issues compiling this branch as well. Full error: https://pastebin.com/6mc9eRed

Mostly errors like this:
```
  CXXLD    libbitcoinconsensus.la

Undefined symbols for architecture x86_64:
  "ChaCha20::ChaCha20(unsigned char const*, unsigned long)", referenced from:
```

Command: ` ./autogen.sh && ./configure && make && make check` 

I didn't git clean or anything, probably had another PR branch up in my local repo before attempting to compile, looked good up until  `libbitcoinconsensus.la`

System info:

```
--> uname -a
Darwin   18.7.0 Darwin Kernel Version 18.7.0: Tue Aug 20 16:57:14 PDT 2019; root:xnu-4903.271.2~2/RELEASE_X86_64 x86_64
--> gcc -v
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple clang version 11.0.0 (clang-1100.0.33.8)
Target: x86_64-apple-darwin18.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

pinheadmz commented:
@naumenkogs the new test I think has the wrong permissions:

```
-rwxr-xr-x    1 matthewzipkin  staff    5718 Feb  3 10:10 p2p_disconnect_ban.py
-rwxr-xr-x    1 matthewzipkin  staff    3833 Dec  3 15:09 p2p_dos_header_tree.py
-rw-r--r--    1 matthewzipkin  staff   28920 Feb 10 09:34 p2p_erlay.py
-rwxr-xr-x    1 matthewzipkin  staff    5741 Feb  3 10:10 p2p_eviction.py
-rwxr-xr-x    1 matthewzipkin  staff    5185 Feb  3 10:10 p2p_feefilter.py
```

```
$ test/functional/p2p_erlay.py
-bash: test/functional/p2p_erlay.py: Permission denied
```

ariard commented on src/net_processing.cpp:
ce83d3fb

I think we should have this discussion about just throwing away transactions under some `SKETCH_MAX_SIZE`. Sort them by feerate, keep the ones above the mark, discard the others.

We're already doing the same assumptions if a a transaction is under mempool min fee, we won't relay it further. That can be the same with reconciliation, we don't allocate sketch resources for low-fee transactions. 

That said, I feel it should be part of a wider conversation about how we price better network resources while still flowing through the network what is "near confirmation".

ariard reviewed (commented):
Thanks for working on this, the scope of topics covered is really wide and interesting : privacy attacks, 
network resources usage, latency, ...

What I'm still not able to reason on easily is the selection of the tx-announcement protocol,
which AFAIU is function of both the peer types (strategic outbound, non-strategic outbound, inbound),
the transaction origin considered (initial-broadcast, relay-from-flooding, relay-from-initiated-reconciliation,
relay-from-accepted-reconciliation) and lastly the sketch processing limits.

I think it's more a work of tightening better the bip/paper to the implementation comments, and
mark the divergences from them if they are some. 

W.r.t to encapsulation, it could be at least to move new messages handlers in their own methods the
same way it has been done for BIP157 (`ProcessReconcilDiff`, `ProcessSketch`, ...).

jonatack commented:
> @naumenkogs the new test I think has the wrong permissions:
> -bash: test/functional/p2p_erlay.py: Permission denied

Confirming @pinheadmz' comment, I needed to run `chmod 755 test/functional/p2p_erlay.py` on the file. The test then fails with `Assertion failed: detected inconsistent lock order for 'cs_main' in net_processing.cpp:2933 (in thread 'msghand')`

<details><summary>traceback</summary><p>

```
Traceback (most recent call last):
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/p2p_erlay.py", line 583, in <module>
    ReconciliationTest().main()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 149, in main
    exit_code = self.shutdown()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 278, in shutdown
    self.stop_nodes()
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 525, in stop_nodes
    node.stop_node(wait=wait, wait_until_stopped=False)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_node.py", line 334, in stop_node
    raise AssertionError("Unexpected stderr {} != {}".format(stderr, expected_stderr))
AssertionError: Unexpected stderr Assertion failed: detected inconsistent lock order for 'cs_main' in net_processing.cpp:2933 (in thread 'msghand'), details in debug log. != 
[node 0] Cleaning up leftover process
```
</p></details>

naumenkogs committed:
Squashed 'src/minisketch/' content from commit a570de261

git-subtree-dir: src/minisketch
git-subtree-split: a570de26152a0ec224c38f8157f796030cb0bd20

naumenkogs committed:
Merge commit '156004af80c20f0967ac867ca900086bcc17d37f' as 'src/minisketch'

pinheadmz commented:
@jonasnick I `chmod a+x` and the test ran and passed.

jonatack commented:
> the test ran and passed.

Building with `--enable-debug`? The lock order errors seem pertinent to these changes. I even rebuilt again a few hours later yesterday after updating my depends; same thing. It's odd if no one else is seeing them.

jonatack commented:
I suppose if no one is seeing the issues, I'll wait until the next push here and look at what the CI says.

naumenkogs committed:
Process incoming reconciliation request

Upon receiving a reconciliation request, a node stores
it and schedules a response.
Do not respond to a request right away as it would enable
a DoS attack and allow monitoring of transaction a node has.
Instead, respond to all reconciliation requests at a same time
after a small delay.

naumenkogs committed:
Add helper to compute reconciliation tx short id

Short ids are used to compute reconciliation sketches.

amitiuttarwar commented on src/net_processing.cpp:
```diff
@@ -2806,6 +2820,61 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+            flood_to = true;
+        }
+
+        uint64_t local_salt = peer->m_local_recon_salt;
```
this salt calculation (and probably also the flood_to logic) seems like an example of code that can be extracted into the reconciliation module rather than added to `ProcessMessage`.

it would make sense to me for the flow to be something like: 
- `ProcessMessage` does internal state checks (`m_tx_relay` exists, wtxid is enabled)
- extract the message from `vRecv`
- pass it along into the module to handle 
- module returns saying "no-op" or "here is a ReconState object" 
- if it exists, `ProcessMessage` stores the object on the peer

I suspect we could go even further, but this would be a start. Ideally the module would abstract this nuanced erlay specific logic away from net processing, even if its per-peer. The txrequest module is a good example for this, eg. `PeerInfo` struct and how its applied within the module to make choices, and the decisions are surfaced to `net_processing`.

naumenkogs replied:
Yeah, I'm planning to do something similar in the next couple days. That's why I marked the PR draft for now :)

Thank you for the design suggestion!

amitiuttarwar commented on src/net_processing.cpp:
```diff
@@ -914,6 +929,28 @@ void PeerManagerImpl::PushNodeVersion(CNode& pnode, int64_t nTime)
+    m_connman.ForEachNode([this, &result, &skip_node](const CNode* pnode) {
+        if (!pnode->m_tx_relay) return;
+        if (pnode->GetId() == skip_node.GetId()) return;
+        if (!pnode->IsFullOutboundConn() && !pnode->IsManualConn()) return;
```
I find (!A && !B) to be very confusing. Again, would be good to extract into its own function with switch statement.

amitiuttarwar commented on src/net_processing.cpp:
```diff
@@ -914,6 +929,28 @@ void PeerManagerImpl::PushNodeVersion(CNode& pnode, int64_t nTime)
     }
 }
 
+size_t PeerManagerImpl::GetFloodingOutboundsCount(const CNode& skip_node) const
```
instead of recalculating whenever we get a `SENDRECON` message, could this logic be significantly simplified by just storing a count that gets updated when nodes do something to change their state? the characteristics being checked are all established early in the connection (connection type, m_tx_relay struct, flood_to being set) 

I think this should remove the need for the "skip node" logic in this function, but if not, the node being processed can be checked and then subtract one if needed.

amitiuttarwar reviewed (commented):
making my way through the first couple commits

naumenkogs commented:
An update here.
Based on some suggestions, I moved reconciliation to a separate module.
I suggest doing some initial review of that in [the PR in my repo](https://github.com/naumenkogs/bitcoin/pull/2). Could you folks help to evaluate my new approach? @jnewbery @amitiuttarwar @sipa @jonatack (everyone else is welcome too)

This way we could save the review efforts of those contributors who are less concerned about the modularity? First do that, then proceed to other questions.

Once that is done, I'll get back here and welcome full review again.

amitiuttarwar commented:
@naumenkogs I took a first look and I see that most of the comments I've left here are still applicable. I'm happy to do a deeper dive on the branch one you've addressed the outstanding comments.

fanquake commented on src/net_processing.cpp:
```diff
@@ -2781,6 +2933,65 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+        static const auto RECON_SALT_HASHER = TaggedHash(RECON_STATIC_SALT);
+        uint256 full_salt = (CHashWriter(RECON_SALT_HASHER) << salt1 << salt2).GetSHA256();
+
+        peer->m_recon_state = MakeUnique<ReconState>(recon_requestor, recon_responder, flood_to, full_salt.GetUint64(0), full_salt.GetUint64(1));
```
Please [use `std::make_unique` in new code](https://github.com/bitcoin/bitcoin/blob/master/src/util/memory.h#L13).

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -1665,17 +1680,19 @@ bool static AlreadyHaveBlock(const uint256& block_hash) EXCLUSIVE_LOCKS_REQUIRED
-            pnode->PushTxInventory(txid);
+            // Reconciliations are not supported for non-wtxid peers,
+            // so we always use flooding.
+            pnode->PushTxInventory(txid, true);
```
I know it's more complicated logic, but think it's necessary.
m_flood inside peer's ReconState is one factor, but it's not sufficient, because we can't flood *all* transactions to them.

If we did so, there will be too much flooding. A non-reachable node always sets m_flood for its 8 outbound peers to true. In that case, it will flood all transactions it gets for them, too much bandwidth.

I suggest flooding only transactions received via inbound flooding, so this is never the case for non-reachable nodes.

I'm open to suggestions on other ways how to limit flooding, but I think this one makes the most sense.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -2795,6 +2808,61 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+        peer->m_recon_state = MakeUnique<ReconState>(recon_requestor, recon_responder, flood_to, full_salt.GetUint64(0), full_salt.GetUint64(1));
+
+        // Reconcile with all outbound peers supporting reconciliation (even if we flood to them),
+        // to not miss transactions they have for us but won't flood.
```
"strategic" is not a bidirectional link property, it's decided on every peer locally.
We might flood stuff to them, but they won't flood to us (because they only flood outbound). 

In that case, we will never get their transactions, if we don't reconcile.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -1017,7 +1024,15 @@ void PeerManagerImpl::ReattemptInitialBroadcast(CScheduler& scheduler) const
+            // flooding non-reachable nodes. Instead, we relay them via reconciliation,
+            // in which case a receiver can't distinguish them from transactions we reconciled
+            // with some other peer.
+            RelayTransaction(txid, tx->GetWitnessHash(), m_connman, false);
```
>Non-reachable nodes may open connections to outbound not supporting reconciliations, thus relying on flooding for transaction announcements, at least until Erlay is well-deployed.

Sure. But for reconciliation conns, the current policy is to flood only what was received via inbound. So if we rebroadcast to a reconciling peer via flooding, they will know it's rebroadcast, because non-reachable never receive any stuff via inbound (by definition)

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -4504,11 +4528,11 @@ bool PeerManager::SendMessages(CNode* pto)
-                        pto->m_tx_relay->nNextInvSend = std::chrono::microseconds{m_connman.PoissonNextSendInbound(count_microseconds(current_time), INVENTORY_BROADCAST_INTERVAL)};
-                    } else {
-                        // Use half the delay for outbound peers, as there is less privacy concern for them.
+                    if (peer->m_recon_state || !pto->IsInboundConn()) {
```
>It seems like an inbound peer that constantly tries to reconcile with us could be a more effective spy than before this change, but maybe I'm missing something.

We don't respond to reconciliations right away, there is a shared timer for those responses. That's why I thought it's fine to reduce the delay here.

naumenkogs commented on src/net_processing.cpp:
```diff
@@ -3358,7 +3375,13 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
+
+            // Flood those transactions which were received either via flooding, or inbound reconciliation,
+            // but NOT via outbound reconciliation. Flooding then is mainly used for initial propagation
+            // of new transactions across a network of reachable nodes quickly.
```
>IIUC, this comment correctly, you mean that a transaction discovered through an accepted-reconciliation will be marked for reconciliation for its future announcements, no matter the peer type (strategic outbound/non-strategic outbound/inbound).

Correct if you mean accepted=outbound (see confusion below)

> I don't get the rational of such approach, propagation of such transactions could be accelerated if we were flooding them for our outbound peers.

Consider a non-reachable peer with 8 reconciliation-supporting outbound peers. They will be all marked strategic. If we do what you say, we will flood transactions to all of them. I suggest limiting it: since a non-reachable node received it from an outbound reconciliation (that's the only way it could have received a tx), it won't flood them to anyone.

>Also you may consider "inbound reconciliation" -> "initiated reconciliation", "outbound reconciliation" -> "accepted reconciliation". Even if reconciliation roles are decided in function of our peer selection, they should be clearly dissociated as orthogonal.

ACK, although I think "inbound" is not initiated, but it's accepted.

naumenkogs commented:
Closing in favor of #21515. I believe I addressed most of the comments from here.

naumenkogs commented:
#21515 is ready for review now
